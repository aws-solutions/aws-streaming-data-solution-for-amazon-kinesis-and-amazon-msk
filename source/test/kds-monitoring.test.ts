/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */


import * as cdk  from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { DataStreamMonitoring } from '../lib/kds-monitoring';

let stack: cdk.Stack;

beforeAll(() => {
    const app = new cdk.App();
    stack = new cdk.Stack(app, 'TestStack');
});

describe('KDS only monitoring', () => {
    beforeAll(() => {
        new DataStreamMonitoring(stack, 'TestAlarms1', {
            streamName: 'test-stream'
        });
    });

    test('creates IteratorAgeMilliseconds alarm', () => {
        Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
            MetricName: 'GetRecords.IteratorAgeMilliseconds',
            Namespace: 'AWS/Kinesis',
            ComparisonOperator: 'GreaterThanOrEqualToThreshold',
            Statistic: 'Maximum',
            EvaluationPeriods: 1,
            Period: 60,
            Threshold: 60000,
            TreatMissingData: 'breaching',
            Dimensions: [{
                Name: 'StreamName',
                Value: 'test-stream'
            }]
        });
    });

    test('creates ReadProvisionedThroughputExceeded alarm', () => {
        Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
            MetricName: 'ReadProvisionedThroughputExceeded',
            Namespace: 'AWS/Kinesis',
            ComparisonOperator: 'GreaterThanOrEqualToThreshold',
            Statistic: 'Average',
            EvaluationPeriods: 1,
            Period: 60,
            Threshold: 0.01,
            TreatMissingData: 'breaching',
            Dimensions: [{
                Name: 'StreamName',
                Value: 'test-stream'
            }]
        });
    });

    test('creates WriteProvisionedThroughputExceeded alarm', () => {
        Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
            MetricName: 'WriteProvisionedThroughputExceeded',
            Namespace: 'AWS/Kinesis',
            ComparisonOperator: 'GreaterThanOrEqualToThreshold',
            Statistic: 'Average',
            EvaluationPeriods: 1,
            Period: 60,
            Threshold: 0.01,
            TreatMissingData: 'breaching',
            Dimensions: [{
                Name: 'StreamName',
                Value: 'test-stream'
            }]
        });
    });

    test('creates PutRecord alarm', () => {
        Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
            MetricName: 'PutRecord.Success',
            Namespace: 'AWS/Kinesis',
            ComparisonOperator: 'LessThanOrEqualToThreshold',
            Statistic: 'Average',
            EvaluationPeriods: 1,
            Period: 60,
            Threshold: 0.95,
            TreatMissingData: 'notBreaching',
            Dimensions: [{
                Name: 'StreamName',
                Value: 'test-stream'
            }]
        });
    });

    test('creates PutRecords alarm', () => {
        Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
            MetricName: 'PutRecords.Success',
            Namespace: 'AWS/Kinesis',
            ComparisonOperator: 'LessThanOrEqualToThreshold',
            Statistic: 'Average',
            EvaluationPeriods: 1,
            Period: 60,
            Threshold: 0.95,
            TreatMissingData: 'notBreaching',
            Dimensions: [{
                Name: 'StreamName',
                Value: 'test-stream'
            }]
        });
    });

    test('creates GetRecords alarm', () => {
        Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
            MetricName: 'GetRecords.Success',
            Namespace: 'AWS/Kinesis',
            ComparisonOperator: 'LessThanOrEqualToThreshold',
            Statistic: 'Average',
            EvaluationPeriods: 1,
            Period: 60,
            Threshold: 0.98,
            TreatMissingData: 'breaching',
            Dimensions: [{
                Name: 'StreamName',
                Value: 'test-stream'
            }]
        });
    });
});

describe('KDS + Lambda monitoring', () => {
    test('creates a dashboard for KDS stream and Lambda function', () => {
        new DataStreamMonitoring(stack, 'TestAlarms2', {
            streamName: 'test-stream',
            lambdaFunctionName: 'test-function'
        });

    });
});
