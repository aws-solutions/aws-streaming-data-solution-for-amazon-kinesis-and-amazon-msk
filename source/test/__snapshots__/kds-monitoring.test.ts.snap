// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KDS + Lambda monitoring creates a dashboard for KDS stream and Lambda function 1`] = `
Object {
  "Resources": Object {
    "TestAlarms1Dashboard5C0CF01C": Object {
      "Properties": Object {
        "DashboardBody": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"widgets\\":[{\\"type\\":\\"text\\",\\"width\\":24,\\"height\\":1,\\"x\\":0,\\"y\\":0,\\"properties\\":{\\"markdown\\":\\"\\\\n# Kinesis Data Stream Metrics\\\\n\\"}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":0,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Get records iterator age (Milliseconds)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1IteratorAgeAlarm554AEEA8",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":6,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Read throughput exceeded (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1ReadProvisionedAlarm869B96F0",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":12,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Write throughput exceeded (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1WriteProvisionedAlarm86317597",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":18,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Put record success (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1PutRecordAlarm88A04D43",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":0,\\"y\\":7,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Put records success (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1PutRecordsAlarmF586CCF5",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":6,\\"y\\":7,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Get records success (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1GetRecordsAlarm68C8EF5F",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}}]}",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "TestAlarms1GetRecordsAlarm68C8EF5F": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "GetRecords.Success",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.98,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1IteratorAgeAlarm554AEEA8": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "GetRecords.IteratorAgeMilliseconds",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Maximum",
        "Threshold": 60000,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1PutRecordAlarm88A04D43": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PutRecord.Success",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.95,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1PutRecordsAlarmF586CCF5": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PutRecords.Success",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.95,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1ReadProvisionedAlarm869B96F0": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ReadProvisionedThroughputExceeded",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.01,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1WriteProvisionedAlarm86317597": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "WriteProvisionedThroughputExceeded",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.01,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms2Dashboard42915699": Object {
      "Properties": Object {
        "DashboardBody": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"widgets\\":[{\\"type\\":\\"text\\",\\"width\\":24,\\"height\\":1,\\"x\\":0,\\"y\\":0,\\"properties\\":{\\"markdown\\":\\"\\\\n# Kinesis Data Stream Metrics\\\\n\\"}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":0,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Get records iterator age (Milliseconds)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms2IteratorAgeAlarm0147B6CD",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":6,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Read throughput exceeded (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms2ReadProvisionedAlarm7259474E",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":12,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Write throughput exceeded (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms2WriteProvisionedAlarmB9E955C9",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":18,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Put record success (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms2PutRecordAlarmDCD19290",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":0,\\"y\\":7,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Put records success (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms2PutRecordsAlarm8C41B160",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":6,\\"y\\":7,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Get records success (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms2GetRecordsAlarm362BC702",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"text\\",\\"width\\":24,\\"height\\":1,\\"x\\":0,\\"y\\":13,\\"properties\\":{\\"markdown\\":\\"\\\\n# Lambda Metrics\\\\n\\"}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":0,\\"y\\":14,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Invocations\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/Lambda\\",\\"Invocations\\",\\"FunctionName\\",\\"test-function\\",\\"Resource\\",\\"test-function\\",{\\"period\\":60,\\"stat\\":\\"Sum\\"}]],\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":6,\\"y\\":14,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Duration\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/Lambda\\",\\"Duration\\",\\"FunctionName\\",\\"test-function\\",\\"Resource\\",\\"test-function\\",{\\"period\\":60,\\"stat\\":\\"Minimum\\"}],[\\"AWS/Lambda\\",\\"Duration\\",\\"FunctionName\\",\\"test-function\\",\\"Resource\\",\\"test-function\\",{\\"period\\":60}],[\\"AWS/Lambda\\",\\"Duration\\",\\"FunctionName\\",\\"test-function\\",\\"Resource\\",\\"test-function\\",{\\"period\\":60,\\"stat\\":\\"Maximum\\"}]],\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":12,\\"y\\":14,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Error count and success rate (%)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/Lambda\\",\\"Errors\\",\\"FunctionName\\",\\"test-function\\",\\"Resource\\",\\"test-function\\",{\\"color\\":\\"#d62728\\",\\"period\\":60,\\"stat\\":\\"Maximum\\",\\"id\\":\\"errors\\"}],[{\\"label\\":\\"Success rate (%)\\",\\"color\\":\\"#2ca02c\\",\\"expression\\":\\"100 - 100 * errors / MAX([errors, invocations])\\",\\"period\\":60,\\"yAxis\\":\\"right\\"}],[\\"AWS/Lambda\\",\\"Invocations\\",\\"FunctionName\\",\\"test-function\\",\\"Resource\\",\\"test-function\\",{\\"label\\":\\"\\",\\"period\\":60,\\"stat\\":\\"Sum\\",\\"visible\\":false,\\"id\\":\\"invocations\\"}]],\\"yAxis\\":{\\"right\\":{\\"max\\":100}}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":18,\\"y\\":14,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Throttles\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/Lambda\\",\\"Throttles\\",\\"FunctionName\\",\\"test-function\\",\\"Resource\\",\\"test-function\\",{\\"period\\":60,\\"stat\\":\\"Sum\\"}]],\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":0,\\"y\\":20,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"IteratorAge\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/Lambda\\",\\"IteratorAge\\",\\"FunctionName\\",\\"test-function\\",\\"Resource\\",\\"test-function\\",{\\"period\\":60,\\"stat\\":\\"Maximum\\"}]],\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":6,\\"y\\":20,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Concurrent executions\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/Lambda\\",\\"ConcurrentExecutions\\",\\"FunctionName\\",\\"test-function\\",\\"Resource\\",\\"test-function\\",{\\"period\\":60,\\"stat\\":\\"Maximum\\"}]],\\"yAxis\\":{}}}]}",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "TestAlarms2GetRecordsAlarm362BC702": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "GetRecords.Success",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.98,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms2IteratorAgeAlarm0147B6CD": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "GetRecords.IteratorAgeMilliseconds",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Maximum",
        "Threshold": 60000,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms2PutRecordAlarmDCD19290": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PutRecord.Success",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.95,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms2PutRecordsAlarm8C41B160": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PutRecords.Success",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.95,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms2ReadProvisionedAlarm7259474E": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ReadProvisionedThroughputExceeded",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.01,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms2WriteProvisionedAlarmB9E955C9": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "WriteProvisionedThroughputExceeded",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.01,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;

exports[`KDS only monitoring creates a dashboard for KDS stream 1`] = `
Object {
  "Resources": Object {
    "TestAlarms1Dashboard5C0CF01C": Object {
      "Properties": Object {
        "DashboardBody": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"widgets\\":[{\\"type\\":\\"text\\",\\"width\\":24,\\"height\\":1,\\"x\\":0,\\"y\\":0,\\"properties\\":{\\"markdown\\":\\"\\\\n# Kinesis Data Stream Metrics\\\\n\\"}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":0,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Get records iterator age (Milliseconds)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1IteratorAgeAlarm554AEEA8",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":6,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Read throughput exceeded (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1ReadProvisionedAlarm869B96F0",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":12,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Write throughput exceeded (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1WriteProvisionedAlarm86317597",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":18,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Put record success (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1PutRecordAlarm88A04D43",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":0,\\"y\\":7,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Put records success (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1PutRecordsAlarmF586CCF5",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":6,\\"x\\":6,\\"y\\":7,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Get records success (Percent)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarms1GetRecordsAlarm68C8EF5F",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}}]}",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "TestAlarms1GetRecordsAlarm68C8EF5F": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "GetRecords.Success",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.98,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1IteratorAgeAlarm554AEEA8": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "GetRecords.IteratorAgeMilliseconds",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Maximum",
        "Threshold": 60000,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1PutRecordAlarm88A04D43": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PutRecord.Success",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.95,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1PutRecordsAlarmF586CCF5": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PutRecords.Success",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.95,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1ReadProvisionedAlarm869B96F0": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ReadProvisionedThroughputExceeded",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.01,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarms1WriteProvisionedAlarm86317597": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StreamName",
            "Value": "test-stream",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "WriteProvisionedThroughputExceeded",
        "Namespace": "AWS/Kinesis",
        "Period": 60,
        "Statistic": "Average",
        "Threshold": 0.01,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;
